File: 684-redundant-connection/redundant-connection.java
class Solution {
  class DisjointSet {
    List<Integer> parent;
    List<Integer> size;

    public DisjointSet(int n) {
      parent = new ArrayList<>();
      size = new ArrayList<>();

      for (int i = 0; i < n; i++) {
        parent.add(i);
        size.add(1);
      }

    }

    public int find(int node){
      if(node == parent.get(node)){
        return node;
      }

      parent.set(node , find(parent.get(node)));
      return parent.get(node);
    }

    public void union(int u , int v){
      int pu = find(u);
      int pv = find(v);

      if(pu == pv) {
        return;
      }

      int sizeu = size.get(pu);
      int sizev = size.get(pv);

      if(sizeu > sizev){
        parent.set(pv , pu);
        size.set(pu , sizeu + sizev);
      } else{
        parent.set(pu , pv);
        size.set(pv , sizeu + sizev);
      }
    }

  }

  public int[] findRedundantConnection(int[][] edges) {
    DisjointSet ds = new DisjointSet(edges.length + 1);
    int[] pair = new int[]{};
    for(int[] edge : edges){
      int u = edge[0];
      int v = edge[1];

      if(ds.find(u) == ds.find(v)){
        pair = edge;
      } else{
        ds.union(u , v);
      }
    }

    return pair;

    
  }
}

File: 684-redundant-connection/redundant-connection.py
class DSU:

    def __init__(self , n):
        self.size = [1 for _ in range(n)]
        self.parent = [i for i in range(n)]
    
    def find(self, node):
        if self.parent[node] != node:
            self.parent[node] = self.find(self.parent[node])
        
        return self.parent[node]
    
    def union(self, u , v):
        pu = self.find(u)
        pv = self.find(v)

        if pu == pv: return

        sizeu = self.size[pu]
        sizev = self.size[pv]

        if sizeu > sizev:
            self.size[pu] = sizeu + sizev
            self.parent[pv] = pu
        else:
            self.size[pv] = sizeu + sizev
            self.parent[pu] = pv  

class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        dsu = DSU(n+1)

        ans = []

        for u,v in edges:
            if dsu.find(u) == dsu.find(v):
                ans = [u,v]
            else:
                dsu.union(u,v)
        
        return ans